Infrastructure: Add MSYS2 workflow for compiling 64bit Mudlet with Qt6 for Windows

jmckisson opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions

This is intended to replace Appveyor for the Windows CI/CB process. It successfully creates -testing versions however the ptr and release version building needs to be tested by the Admin team.

#### Motivation for adding to Mudlet

#### Other info (issues closed, discussion etc)

Appveyor still attempts to run so the integration probably needs to be removed by the admin staff.

jmckisson committed:
add github workflow for win32 compile

jmckisson committed:
fix github workspace test

jmckisson committed:
make github workspace path a unix path

jmckisson committed:
add packaging step

jmckisson committed:
remove appveyor and build-mudlet-win workflow

jmckisson committed:
change matri.os to runner.os

jmckisson committed:
change runner.os check to Windows

jmckisson committed:
fix bad indent

jmckisson committed:
fix script invokation path

jmckisson committed:
properly execute the deploy powershell script

jmckisson committed:
fix string and boolean values

jmckisson committed:
fix/debug check for ptb and testing builds

jmckisson committed:
setup github env for upload directory, testing release only

jmckisson committed:
guard non-windows actions

jmckisson committed:
dont produce zip file in package script

jmckisson committed:
name build steps

jmckisson committed:
switch to 64bit and qt6

jmckisson committed:
change repo owner back to Mudlet

jmckisson committed:
change to qt6keychain

jmckisson committed:
no longer need core5compat

jmckisson committed:
apparently we still need core5compat

jmckisson committed:
fikx keychain include in dlgConnectionProfiles header

jmckisson committed:
apparently --release flag for windeployqt is broken

jmckisson committed:
pr 7214 fix edbee crash in Qt6

jmckisson committed:
resolve merge conflicts including submodule update

jmckisson committed:
fix upstream edbee lib

jmckisson committed:
clean up build scripts to no longer require install or build type

jmckisson committed:
need define for MINGW_BASE_DIR

jmckisson committed:
convert powershell deploy script to bash

jmckisson committed:
need BUILD_CONFIG to specify proper directory

jmckisson committed:
fix script error, cleanup

jmckisson committed:
add 7zip

jmckisson committed:
fix version extraction

jmckisson committed:
Merge branch 'development' into development

vadi2 commented on .github/workflows/build-mudlet.yml:
```diff
@@ -56,15 +55,43 @@ jobs:
+      shell: msys2 {0}
+      run: $GITHUB_WORKSPACE/CI/package-mudlet-for-windows.sh
+        
+    - name: (Windows) Deploy
```
Given that this part of the build is purely about Windows and then the rest of the build is macos/Linux, perhaps this should be a separate workflow? What do you reckon? There doesn't seem to be any re-use happening here and it feels out of place.

jmckisson replied:
Moving to a separate windows only workflow is easy enough, I had it that way originally before incorporating it into the main one.

vadi2 commented on CI/build-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,133 @@
+fi
+
+# The updater is not helpful in this (build it yourself) environment
+export WITH_UPDATER="NO"
```
Is this bit still needed?

jmckisson replied:
Does the updater need to be enabled for release builds?

vadi2 replied:
Yeah definitely.

SlySven replied:
But fix https://github.com/Mudlet/Mudlet/pull/7217#pullrequestreview-2041248681 first - or things will get messy/borked if we start injecting 64-Bit updates into those users who are actually using 32-bit Windows - fair enough it won't be an issue for release builds until Vadi and the rest of us Core Dev team get our act together and get one out but PTBs will adopt this within 24 hours of it being merged into the `development` branch code. Note, of course, that we will have to go through the PTB and release execution paths with a fine-tooth comb because ***they are hard, if not, impossible, to test completely before being used for* real!**

vadi2 commented on CI/build-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,133 @@
+  mingw32-make -k
+fi
+
+echo " ... make finished - has it been successful?"
```
Mm, what is this trying to say?

jmckisson replied:
Remnant from SlySven's scripts.. Surely not needed anymore since it will continue automatically now

jmckisson committed:
update copyright, add exit codes, fix deploy PTB logic

jmckisson committed:
Merge branch 'development' of github.com:jmckisson/Mudlet into development

jmckisson committed:
add win32 build

jmckisson committed:
32 bit needs own qt keychain

jmckisson committed:
fix package script error

jmckisson committed:
differentiate between 32 and 64 bit builds for artifact

vadi2 commented on CI/deploy-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,331 @@
+  fi
+
+  echo "=== Moving files to upload directory ==="
+  mv "$PACKAGE_DIR/$uploadFilename" "$uploadDir/"
```
Just a heads up, Github always zips artifacts, so it is double zipped right now.

vadi2 reviewed (commented):
Both 32bit and 64bit versions seem to work well, great stuff! 👍

SlySven commented on .github/workflows/build-mudlet.yml:
```diff
@@ -40,11 +40,13 @@ jobs:
-#             qt: '5.14.2'
+          - os: windows-2019
+            buildname: 'windows64'
+            triplet: x86_64-w64-mingw32
```
Not sure about this and the other triplet - have these values been plucked out of the air or do they actually represent valid values to use?

jmckisson replied:
So I think the triplets aren't needed for the msys builds as they reference something vcpkg uses?

jmckisson committed:
remove verbosity from echo

SlySven commented on CI/deploy-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,331 @@
+  if [[ "$PublicTestBuild" == "true" ]]; then
+    # Allow public test builds to be installed side by side with the release builds by renaming the app
+    # No dots in the <id>: Guidelines by Squirrel
+    sed -i 's/<id>Mudlet<\/id>/<id>Mudlet-PublicTestBuild<\/id>/' "$NuSpec"
+    sed -i 's/<title>Mudlet<\/title>/<title>Mudlet (Public Test Build)<\/title>/' "$NuSpec"
```
🚨 The names used here MUST be different for the 64-Bit builds (but the ID must still be a valid one - which rules out some obvious candidates) - see:
* https://github.com/Squirrel/Squirrel.Windows/blob/develop/docs/using/nuget-package-metadata.md
* https://github.com/Squirrel/Squirrel.Windows/blob/master/docs/using/naming.md

basically it looked as though `-` is used as a separator (but we seem to be getting away with it in the PTBs, nor `.` or spaces and I've found the id cannot **end** in numbers or use any sort of brace or bracket either. When I was trying something out in the past I had not changed the 32-bit ones from what **was already in use** but changed the 64-bit ones to use `Mudlet_64_` for the id for the release builds and `Mudlet_64_-PublicTestBuild` for the PTBs.

jmckisson committed:
include updater in release builds

jmckisson committed:
update bitness for squirrel, specify dblsqd attach type, typos

jmckisson committed:
triplet not needed for msys2 build

jmckisson committed:
don't pre-zip the artifact

jmckisson committed:
add -u option to pacman

SlySven commented:
Apropos your recent @jmckisson comments on Discord about builds failing - have you noticed this:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/08d7c2a7-1ae6-4a27-9776-a8dbaab6dba8)

If you aren't using AppVeyor anymore you might want to remove the `.appveyor.yml` and related files from the repository. As it happens I am returning to try and get both 32-bit and 64-bits builds out of an AppVeyor CI/CB process and this time I'll try and get the Windows installer stuff operating (for the PTB and Releases)...

jmckisson commented:
> Apropos your recent @jmckisson comments on Discord about builds failing - have you noticed this: ![image](https://private-user-images.githubusercontent.com/6163092/328320461-08d7c2a7-1ae6-4a27-9776-a8dbaab6dba8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwMzA3MTUsIm5iZiI6MTcxNTAzMDQxNSwicGF0aCI6Ii82MTYzMDkyLzMyODMyMDQ2MS0wOGQ3YzJhNy0xYWU2LTRhMjctOTc3Ni1hOGRiYWFiNmRiYTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDZUMjEyMDE1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTU1OWMxZDNjZjI5Yzk2MDljYTM5N2UxMTg4NjgxOGU2Mjg5YzljOTRiM2U1Mjg2NmUwZjUzYWM2ODEyMDhjNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Bf_LTTzckGM9d3BtSbSWi823sFIkrBQzMB9q_OLm3Fg)
> 
> If you aren't using AppVeyor anymore you might want to remove the `.appveyor.yml` and related files from the repository. As it happens I am returning to try and get both 32-bit and 64-bits builds out of an AppVeyor CI/CB process and this time I'll try and get the Windows installer stuff operating (for the PTB and Releases)...

I did remove the .appveyor.yml but I think appveyor itself may be looking at the repo thru its integration?

jmckisson committed:
add GITHUB_REPO_TAG to env

jmckisson committed:
add tmate action for debugging

jmckisson committed:
run isnt actually failing so add tmate at the beginning

jmckisson committed:
try to add tmate as a separate step

jmckisson committed:
fix accidently deleted checkout step

jmckisson committed:
dont use tmate action multiple times

jmckisson committed:
start tmate detached

jmckisson committed:
only use tmate in the deploy step

jmckisson committed:
need to start tmate before the deploy step separately

jmckisson committed:
dont append zip to the artifact name, remove tmate for now

jmckisson committed:
use unix path for package directory

jmckisson committed:
debug package directory contents

jmckisson committed:
try debug package dir again

jmckisson committed:
remove wildcard from mv

jmckisson committed:
list files in upload dir

jmckisson committed:
copy instead of move

jmckisson committed:
try rsync maybe

jmckisson committed:
revert back to zipping

jmckisson committed:
Try adding / to upload dir path

jmckisson committed:
fix script error

jmckisson committed:
move files not directory

jmckisson committed:
use rsync instead of mv

jmckisson committed:
enable tmate for debugging again

jmckisson committed:
move windows build to separate workflow

jmckisson committed:
try windows style path for artifact

jmckisson committed:
fix unix path for rsync

jmckisson committed:
fix wrong path

jmckisson committed:
clean things up after testing

jmckisson committed:
set sha for build step

vadi2 committed:
Change curl out for invoke-webrequest

vadi2 committed:
Clean up job names and conditions

Windows will be run separately for the time being

vadi2 committed:
Remove extra space from step names

vadi2 commented:
Fixed the 32bit submission request. Need to add ccache back (I'll look around at this) and fix the [mg issue](https://discord.com/channels/283581582550237184/427919962561052673/1237775431768342570). Also need to add [validation](https://github.com/Mudlet/Mudlet/blob/development/CI/appveyor.validate_deployment.ps1) back, this helps prevent manual mistakes when deploying.

vadi2 committed:
Try and see if ccache can be installed

vadi2 committed:
Give job a unique name

vadi2 committed:
Install general ccache instead

Per https://packages.msys2.org/package/ccache?repo=msys&variant=x86_64 it should be available in /usr/bin automatically.

vadi2 committed:
Add cache save/restore

vadi2 committed:
Fix updater to be included in PTBs

vadi2 committed:
Set CCACHE_DIR

vadi2 committed:
Run ccache check in msys2 shell

vadi2 committed:
Set PATH before checking ccache

vadi2 committed:
Undo setting the path, turns out that is unnecessary

vadi2 committed:
Check if ccache exists in mingw64

vadi2 committed:
Stop on first error

Continuing to build past the first error makes the actual error hard to find, because it'll be swamped with a lot of other text after the error that is not relevant

jmckisson committed:
add GITHUB_SCHEDULED_BUILD env tag to build step

jmckisson committed:
add release validation step

jmckisson committed:
name file .zip to see if it shows up in json.php

jmckisson committed:
revert back to when artifacts were pre-zipped and showed up inthe json.php

jmckisson committed:
remove .zip extension from uploadFilename

jmckisson committed:
strip extension from zip files

SlySven commented on CI/deploy-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,340 @@
+  if [[ "$PublicTestBuild" == "true" ]]; then
+    # Allow public test builds to be installed side by side with the release builds by renaming the app
+    # No dots in the <id>: Guidelines by Squirrel
+    sed -i "s/<id>Mudlet<\/id>/<id>Mudlet_${BUILD_BITNESS}_-PublicTestBuild<\/id>/" "$NuSpec"
```
No. You must **not** use a different Id for the 32-bit builds compared to what we were using before - otherwise existing installed Mudlet PTB versions will not see this as an update. For those we were using:
**`Mudlet-PublicTestBuild`** as the id and that is what it must stay as.

jmckisson replied:
Err, I understood that the 32 bit should have remained the same.. I guess i meant to wrap that in a test for MINGW64 but forgot to before I did a commit, oopsie

SlySven commented on CI/deploy-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,340 @@
+    # Allow public test builds to be installed side by side with the release builds by renaming the app
+    # No dots in the <id>: Guidelines by Squirrel
+    sed -i "s/<id>Mudlet<\/id>/<id>Mudlet_${BUILD_BITNESS}_-PublicTestBuild<\/id>/" "$NuSpec"
+    sed -i "s/<title>Mudlet<\/title>/<title>Mudlet_${BUILD_BITNESS} (Public Test Build)<\/title>/" "$NuSpec"
```
The restrictions on the "Title" are not so restricted as for the Id and I think **`Mudlet x${BUILD_BITNESS} (Public Test Build)`** or **`Mudlet x86 (Public Test Build)`** & **`Mudlet x64 (Public Test Build)`** would be more readable in the Windows Installed Applications list. And give a hint to 64 Bit Windows users that there might be an alternative for them.

SlySven commented on CI/deploy-mudlet-for-windows.sh:
```diff
@@ -0,0 +1,340 @@
+    sed -i "s/<id>Mudlet<\/id>/<id>Mudlet_${BUILD_BITNESS}_-PublicTestBuild<\/id>/" "$NuSpec"
+    sed -i "s/<title>Mudlet<\/title>/<title>Mudlet_${BUILD_BITNESS} (Public Test Build)<\/title>/" "$NuSpec"
+  else
+    sed -i "s/<id>Mudlet<\/id>/<id>Mudlet_${BUILD_BITNESS}_<\/id>/" "$NuSpec"
+    sed -i "s/<title>Mudlet<\/title>/<title>Mudlet_${BUILD_BITNESS}<\/title>/" "$NuSpec"
```
Same things as the last couple of comments. The id must remain as **`Mudlet`** for the 32 bit release version but should be different for the NEW 64 bit release - and the suggested Id for that is **`Mudlet_64_`** as any form of brackets or braces or an ending with numbers seems to be a problem and dots just borks things right up.

Title can be like I suggest in the previous comments just without the `(Public Test Build)` obviously.

jmckisson committed:
fix squirrel id for PTB and bitness

jmckisson committed:
revert back to directory upload

jmckisson committed:
set unzip=0 and dont specify extension on uploadFilename

jmckisson committed:
add unziop variable to moveToUploadDir func

jmckisson committed:
fix double quoting

jmckisson committed:
change QtConcurrent:run invocation to a lambda

jmckisson committed:
remove unneeded 7zip

jmckisson committed:
add nuget to PATH and tag 64 bit release build

jmckisson committed:
install nodejs for npm

jmckisson committed:
uncomment some dblsql calls

jmckisson committed:
set installer file_name to windows-32 or -64, remove bitness specification from dblsqd calls

jmckisson committed:
Merge branch 'development' into development

vadi2 reviewed (approved):
Lets get this merged to see how test and PTB builds work now. `appveyor.yml` is gone for now, but will be back with https://github.com/Mudlet/Mudlet/pull/7226.
